File: src/main/java/io/github/toberocat/guiengine/function/call/EditComponentFunction.java
package io.github.toberocat.guiengine.function.call;

import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.JsonDeserializer;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.node.TextNode;
import io.github.toberocat.guiengine.GuiEngineApi;
import io.github.toberocat.guiengine.context.GuiContext;
import io.github.toberocat.guiengine.function.GuiFunction;
import io.github.toberocat.guiengine.utils.JsonUtils;
import io.github.toberocat.guiengine.utils.ParserContext;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;

/**
 * Custom GUI function to edit a component's property in the GUI.
 * <p>
 * Created: 29.04.2023
 * Author: Tobias Madlberger (Tobias)
 */
@JsonDeserialize(using = EditComponentFunction.Deserializer.class)
public record EditComponentFunction(@NotNull String target, @NotNull String property,
                                    @NotNull String value) implements GuiFunction {

    public static final String ID = "edit";

    /**
     * Calls the `editXmlComponentById` method using the provided API and context to edit a component's property in the GUI.
     *
     * @param api     The `GuiEngineApi` instance used to interact with the GUI engine.
     * @param context The `GuiContext` instance representing the GUI context in which the component exists.
     */
    @Override
    public void call(@NotNull GuiEngineApi api, @NotNull GuiContext context) {
        context.editXmlComponentById(api, target, xml -> xml.fields().put(property, new TextNode(value)));
    }

    /**
     * Custom deserializer to convert JSON data into an `EditComponentFunction` instance.
     */
    protected static class Deserializer extends JsonDeserializer<EditComponentFunction> {

        @Override
        public EditComponentFunction deserialize(JsonParser p, DeserializationContext ctxt) throws IOException {
            JsonNode node = p.getCodec().readTree(p);
            return new EditComponentFunction(JsonUtils.getOptionalString(new ParserContext(node, null, null), "target").orElseThrow(), JsonUtils.getOptionalString(new ParserContext(node, null, null), "attribute").orElseThrow(), JsonUtils.getOptionalString(new ParserContext(node, null, null), "set-value").orElseThrow());
        }
    }
}

File: src/main/kotlin/io/github/toberocat/guiengine/function/call/EditComponentFunction.kt
package io.github.toberocat.guiengine.function.call

import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.node.TextNode
import io.github.toberocat.guiengine.context.GuiContext
import io.github.toberocat.guiengine.function.GuiFunction
import io.github.toberocat.guiengine.function.GuiFunctionFactory
import io.github.toberocat.guiengine.xml.parsing.ParserContext

/**
 * Custom GUI function to edit a component's attribute in the GUI.
 *
 *
 * Created: 29.04.2023
 * Author: Tobias Madlberger (Tobias)
 */
@JsonDeserialize(using = EditComponentFunction.Deserializer::class)
data class EditComponentFunction(
    val target: String,
    val attribute: String,
    @field:JsonProperty("set-value") @param:JsonProperty("set-value") val value: String
) : GuiFunction {
    override val type = TYPE

    /**
     * Calls the `editXmlComponentById` method using the provided API and context to edit a component's attribute in the GUI.
     *
     * @param context The `GuiContext` instance representing the GUI context in which the component exists.
     */
    override fun call(context: GuiContext) {
        context.editXmlComponentById(target) { it.fields[attribute] = TextNode(value) }
    }

    /**
     * Custom deserializer to convert JSON data into an `EditComponentFunction` instance.
     */
    class Deserializer : GuiFunctionFactory<EditComponentFunction>() {

        override fun build(node: ParserContext) = EditComponentFunction(
            node.string("target").require(TYPE, javaClass),
            node.string("attribute").require(TYPE, javaClass),
            node.string("set-value").require(TYPE, javaClass)
        )
    }

    companion object {
        const val TYPE = "edit"
    }
}
